### DISCUSSION TITLE
Support for W5500 (SPI Ethernet) and WizFi360 (SPI WiFi)

### DISCUSSION THREAD

**Discussion by ghost** [grblHAL/plugins#13] [Ideas]
Hi Team,

Is it possible to add support for the above two SPI devices? It would be great to have these either as separate plugins or as part of existing plugins and would open up a lot of possibilities for STM32 MCU's that have multiple SPI ports but don't natively support Ethernet or WiFi.

I have no idea how this could be coded but maybe someone within this community could build it.

Cheers

**Comment by terjeio** [comment]
Perhaps this [W5500 lwip code](https://github.com/PeterBorisenko/w5500-lwip-freertos) could be used?

I do not own any of these SPI boards so I am afraid I am not able to add the needed code.

**Comment by ghost** [comment]
Hi terjeio,

The W5500 is an easy one to send to you as it is already a plugin module. The WizFi360 is a standalone module that needs to be built onto a plugin board. I can have a few built and sent you if that would be useful?

I also found a few github repositories that seemed to have firmware suitable for the STM32F4xx range of MCU's:

https://github.com/cnoviello/stm32-nucleof4
https://github.com/afiskon/stm32-w5500

https://github.com/WIZnet-WizFi360
https://community.st.com/s/question/0D53W00000xrbsVSAQ/how-to-add-the-cmsis-wifi-drivers-to-a-stm32cubeide-project

**Reply by terjeio** [reply]
> I can have a few built and sent you if that would be useful?

That would be helpful, but I cannot promise any delivery date if you send them - I have far too many projects going on...

Terje Io
Ievegen 120
5617 Rossland
Norway

**Reply by ghost** [reply]
Cool. Do you also need a W5500 SPI Ethernet module? I can send one of those as well once I get the WiFi module built.

Cheers
George

**Reply by terjeio** [reply]
The ethernet module is perhaps the most interesting one, there are wifi modules available, such as the _D1 mini_, that emulates a serial port that can be now. If I understand it correctly the [ESP3D](https://github.com/luc-github/ESP3D) can also be used as a websocket/telnet <> serial bridge with the additional bonus that it makes the WebUI available. When I get time I may also port/modify the grblHAL ESP3D implementation for the Pi Pico W so that it can be used as a bridge, maybe even with MPG/DRO support.

**Reply by ghost** [reply]
Nice. So could the ESP32 then be configured to communicate back to the STM32 via SPI? If you like I can beuild a simple plugin module the uses the ESP32 rather than the WizFi360.
Modules are cheaper and easier to get a hold of as well

**Reply by terjeio** [reply]
> So could the ESP32 then be configured to communicate back to the STM32 via SPI?

Perhaps, but I am not going to start any new projects using the ESP32 - I am moving away from processors that does not support real debugging. They are way to expensive to work with - cheap to buy but costs tons of hours of needless work. I find myself staring at the ESP32 and thinking "why can't I set a breakpoint on this silly bugger...".

The WizFi360 is very likely an ESP8266 with custom firmware, similar to the D1 mini but with the option of using a SPI interface? I have a D1 mini already, but I do not believe it has support for using SPI for interfacing with the lwip stack.

**Comment by ghost** [comment]
No problem. I will build the WizFi360 breakout board and send to you along
with the w5500.

Cheers and have a great day
George

**Comment by terjeio** [comment]
The STM32F4xx and RP2040 drivers has gotten support for W5100S (untested) and W5500 breakouts. SPI pins has to be defined in the map file for it to work - like in [this map](https://github.com/grblHAL/STM32F4xx/blob/c06947d825e3cbd96a4c0b4f204ad30a7e01a2a6/Inc/st_morpho_map.h#L225-L234). Other drivers needs lowlevel [SPI](https://github.com/grblHAL/STM32F4xx/blob/master/Src/spi.c) and [WIZnet](https://github.com/grblHAL/STM32F4xx/blob/master/Src/w5x00_ll_driver.c) support code implemented in a similary way as for the STM32F4xx driver.
